### ISSUE TITLE
Geyser players cannot claim land due to prefix 

### ISSUE THREAD

**Issue by NoobcoreNetwork** [GriefPrevention/GriefPrevention#1360]
So it would seem geyser players with none alphanumeric items in their names cannot claim. Land or be trusted 

### What was expected:
To be able to claim and be trusted 

### Steps to reproduce:
Installed geyser and set prefix of a bedrock player to * for example. Grief prevention says they have been trusted if they do /trust but they can't actually build etc 

GriefPrevention version 16.17.1-2b69191
Latest paper build (updated last night) 

Be awesome to have this fixed. I adore GP for its simplicity and as much as there is alternatives they are overly complicated.

Thankyou! 🏡

**Comment by RoboMWM** [comment]
need more info, logs, etc. I know GP has some sort of regex filtering but I think that only applies for offlineplayer lookups iirc.

**Comment by NoobcoreNetwork** [comment]
> need more info, logs, etc. I know GP has some sort of regex filtering but I think that only applies for offlineplayer lookups iirc.

What exact logs would you like? It happens on a fresh install of grief prevention on any server jar like spigot. Paper etc. Just doesn't seem to register names with none alphanumeric characters. Talking with a few guys who have the same issue.

Cheers!

**Comment by NoobcoreNetwork** [comment]
You can probably test with without geyser by setting the display name to contain * then try to claim. Land or trust someone with * in their display name

**Comment by bigpresh** [comment]
Since GP stores trust as GUIDs and checks them against the interacting player's GUID, I don't see *why* the name should matter (for online players at least).  As long as `Player.getUniqueId()` still returns their UUID I don't see why it shouldn't work.

Do they get the normal message that they can't build here when they try to build in a claim they should be able to but can't?  Does their name appear in `/trustlist` for that claim?

Can they create their own claim, rather than be trusted in another?  Can they then build in that claim?

**Comment by NoobcoreNetwork** [comment]
I'm so confused. Was there an update recently that changed any of this? I just tried again and it worked... but im not joking it 100% wasnt working a week ago, I might have dropped a jar in lately im using 16.17.2b69191 build 

Avoid requesting invalid names from Mojang (old UUID conversion code) (#1092) (de7a137) - @Jikoo

was recent, could this have been what solved my issue?

**Comment by Jikoo** [comment]
No - that only affected some very old conversion code for legacy databases from 1.7.

Possible your usercache was corrupted and player names weren't fetched from it properly, that could result in those players being un-locatable.

**Comment by Jikoo** [comment]
Should also note that Geyser does allow for usernames with spaces in them. Due to how Bukkit parses command arguments, that makes players with names with spaces in them largely incompatible with any plugin's commands. I don't think it's GP's job to create a system allowing names that are completely illegal in Java edition, especially since so many other plugins are affected. It would be "easy" for the Geyser people to write a parser that groups quoted text and then passes it to commands.

Tl;dr of how Geyser/Floodgate work though: offline is whatever, online mode it basically just lies to the server about profile validity. Internally as far as the server is concerned the players are no different.

**Comment by NoobcoreNetwork** [comment]
> Should also note that Geyser does allow for usernames with spaces in them. Due to how Bukkit parses command arguments, that makes players with names with spaces in them largely incompatible with any plugin's commands. I don't think it's GP's job to create a system allowing names that are completely illegal in Java edition, especially since so many other plugins are affected. It would be "easy" for the Geyser people to write a parser that groups quoted text and then passes it to commands.
> 
> Tl;dr of how Geyser/Floodgate work though: offline is whatever, online mode it basically just lies to the server about profile validity. Internally as far as the server is concerned the players are no different.

Floodgate 2 contains the ability to have it auto remove spaces in names which helps mitigate this issue and also using a prefix like a period instead of an astetix seems to solve alot of issues for me command wise.

**Comment by RoboMWM** [comment]
I'll leave this as an issue in case further information points to a root cause, else this would be transferred to a discussion.

**Comment by NoobcoreNetwork** [comment]
> I'll leave this as an issue in case further information points to a root cause, else this would be transferred to a discussion.

Thanks I can't get it to happen again but I'll speak to my friend who it's happening to and see what's he's doing different to me.

**Comment by SlimeCody** [comment]
![image](https://user-images.githubusercontent.com/49579480/120541536-a16e8500-c3af-11eb-91b7-eb826a22e9d6.png)

So I am not getting any error messages for this, if they have ownership of the claim they are able to do everything as expected
![image](https://user-images.githubusercontent.com/49579480/120542062-44270380-c3b0-11eb-9d69-1e599fd831a0.png)

The issue seems to point that the only time it has issues is handling them for the different trusts?

**Comment by Jikoo** [comment]
@SlimeCody Your issue is that you're using a period for your prefix. GP is recognizing that as a permission and is handling it accordingly.

**Comment by jakey1995abc** [comment]
Are there any dev builds of GriefPrevention that have fixed this issue? And if not, does changing the bedrock prefix mean that players on bedrock will lose anything on my server? (mcmmo levels, inventory, their claims, etc...)

**Comment by RoboMWM** [comment]
No idea, but if the uuid doesn't change then I don't see why they would lose data.

**Comment by PenguinSlippers** [comment]
I'm getting this issue, my bedrock players now have . as a prefix instead of the * and since then the /trust is not working for them, it doesn't error but they cant build or anything, how do I fix these please?

**Comment by PenguinSlippers** [comment]
> I'm getting this issue, my bedrock players now have . as a prefix instead of the * and since then the /trust is not working for them, it doesn't error but they cant build or anything, how do I fix these please?

okay I changed the prefix back to * in the floodgate config and it now works again

**Comment by bigpresh** [comment]
To be fair, this is somewhat worded as though it's a bug in GriefPrevention, but it isn't really.

GP's `/trust` command treats any name with a `.` in it as an attempt to trust anyone with a given permission, rather than a particular person, and has done so for a long time.  The fact that Geyser or other random plugins let people have names that aren't valid as Minecraft usernames and happen to clash with this isn't a GP bug.

One workaround which could be considered is change the [code that treats anything containing a `.` as a permission](https://github.com/TechFortress/GriefPrevention/blob/master/src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java#L2952-L2955) to only consider it a permission if the `.` is not at the start of the argument (and a similar change elsewhere in the code that handles untrust and anywhere else we did the same).  That would maintain the ability for people to trust permission nodes like `foo.bar` (and ones without a `.` by surrounding them with `[...]` e.g. `[foo]`) but allow the use of `.` at the beginning to match players with that as a prefix.

The other option would be to deprecate then remove the ability to trust without surrounding with `[...]` which in the long run might be a bit clearer  - then it's "for the vast majority of cases, `/trust playername` will do what you mean (even if there's a weird prefix character), but in the much less common use-case that you want to trust everyone with a given permission, say `/trust [permission.node]` instead - I think that would make things clearer in the long run, but is a change to existing GP behaviour which could trip up people used to being able to just `/trust foo.bar` (but we could perhaps special-case a message that appears if you tried to trust an unrecognised player and had a dot in it, to say e.g. "If you meant to refer to a permission, then say '[permission]' instead")

**Comment by Jikoo** [comment]
I agree that moving to a system where permissions must be bracketed to be trusted would be easier, but we may also want to reconsider all of the trust logic from the ground up at some point. We'll probably see a lot fewer of these style complaints long term if even known permission styled strings are still checked for valid previous players. You know at some point someone's going to have a whole group of people on their offline mode server named `[cool.guy]` or something and this will crop up again.

I'll probably have a look at this again at some point (especially since Geyser now apparently defaults to a period instead of the old asterisk, pain) but it's lower on my priority list so if it's urgent/anyone feels motivated, please feel free to not wait the year it may take me to get there :V

**Comment by Camotoy** [comment]
I'm already looking into this. I'm designing the fix to always check if a player exists before checking if the name has a period, since there *are* somehow official Mojang accounts with periods in their name: https://namemc.com/profile/..Shadow...1

**Comment by cloakfox** [comment]
> I'm already looking into this. I'm designing the fix to always check if a player exists before checking if the name has a period, since there _are_ somehow official Mojang accounts with periods in their name: https://namemc.com/profile/..Shadow...1

Any news on this? We also ran into this issue with our bedrock users.
Edit: I think we'll just merge the PR ourselves.

**Comment by weihao** [comment]
I am running into this issue and I was able to reproduce this error.

Any updates? 

ping @Jikoo

**Comment by Jikoo** [comment]
> I am running into this issue and I was able to reproduce this error.
> 
> Any updates?
> 
> ping @Jikoo

I only have triage rights in this repo.

**Comment by c0wg0d** [comment]
I am having this issue as well.  I'd love to see a fix added to a supported release soon.  bigpresh's suggestion to not treat /trust as a permission if the dot is the first character in the argument string seems like a good solution to me.
